ItsJonQ/g2
# Style System: Enable prop interpolation for styled components

This update allows for styled components (created with `styled.div'...'`) to support [prop interpolation](https://emotion.sh/docs/styled#changing-based-on-props).

This allows for style workflows similar to that of Emotion and Styled Components.

### ✨ Example

In the example below, we're able to adjust the styles of the `FriendlyExample` component through mixins and inline functions.
These functions output a different style rule based on `props` that the component receives.

```jsx
const happyMixin = (props) => {
	return props.happy ? `color: green` : `color: blue`;
};

const niceMixin = (props) => {
	return props.nice ? `background-color: yellow` : `background-color: #ddd`;
};

const FriendlyExample = styled.div`
	color: #aaa;
	${happyMixin};
	${niceMixin};
	font-size: 40px;
	font-weight: bold;
	padding: 20px;
	${(props) => props.title && `font-size: 90px`}
`;

const App = () => {
	return (
		<FriendlyExample happy={false} nice={false} title="Weee">
			Hello
		</FriendlyExample>
	);
};
```

### 🤔 How does this work?

This is achieved by resolving the "tagged template literal" arguments - a combination of `string` and `function` objects that the `styled` component receives (via the backtick).

This strategy is the same as the one from libraries like Styled Components:
https://mxstbr.blog/2016/11/styled-components-magic-explained/

This update handles this compiling mechanic with this dedicated function:

```js
export function compileInterpolatedStyles(interpolatedStyles, props) {
	const resolvedInterpolatedStyles = interpolatedStyles.map((a) =>
		is.function(a) ? a(props) : a,
	);

	const [cssChunk, ...interpolatedCSSRule] = resolvedInterpolatedStyles;

	const rules = [...cssChunk].slice(0, -1);
	const end = [...cssChunk].slice(-1);

	const compiledStyles = rules
		.map((rule, index) => rule + interpolatedCSSRule[index])
		.join('')
		.concat(end);

	return compiledStyles;
}
```

---

### 🕹 Testing

* Play around with/test out the [`styledComponentInterpolation.stories.js` story](http://localhost:6006/iframe.html?id=create-styles-styledcomponent-stringinterpolation--default&viewMode=story)
* Alternatively, dive into an existing story and create a styled component with mixins.
* Ensure it works
* Ensure existing component styles aren't broken (🤣)

---

Resolves: https://github.com/ItsJonQ/g2/issues/82
